File: backend/src/websocket-handler.js
// backend/src/websocket-handler.js
const WebSocket = require("ws");
const EventEmitter = require("events");
const JWTTokenManager = require('./jwt-utils');
const fetch = require('node-fetch');

class CoinbaseWebSocketHandler extends EventEmitter {
  constructor(config) {
    super();

    // Validate required API credentials
    if (!config.apiKey || !config.signingKey) {
      throw new Error('COINBASE_API_KEY and COINBASE_SIGNING_KEY are required for Advanced Trade WebSocket');
    }

    this.config = {
      wsUrl: config.wsUrl || "wss://advanced-trade-ws.coinbase.com",
      cryptoConfig: config.cryptoConfig || null, // null means fetch all pairs dynamically
      cryptos: [], // Will be populated dynamically
      volumeThreshold: config.volumeThreshold || 1.5,
      windowMinutes: config.windowMinutes || 5,
      reconnectDelay: config.reconnectDelay || 1000,
      maxReconnectDelay: config.maxReconnectDelay || 60000,
      apiKey: config.apiKey,
      signingKey: config.signingKey,
    };

    this.ws = null;
    this.reconnectAttempts = 0;
    this.volumeWindows = {};
    this.historicalAvgs = {};
    this.currentTickers = {};

    // Priority system for active trades
    this.priorityPairs = new Set(); // Set of product_ids that need priority updates
    this.priorityUpdateInterval = 1000; // Priority updates every 1 second
    this.lastPriorityUpdate = {}; // Track last update time for priority pairs

    // Initialize JWT token manager for Advanced Trade WebSocket
    this.jwtManager = new JWTTokenManager(this.config.apiKey, this.config.signingKey);
    this.tokenRefreshInterval = null;

    // Initialize data structures for each crypto
    this.config.cryptos.forEach((crypto) => {
      this.volumeWindows[crypto] = [];
      this.historicalAvgs[crypto] = 0;
      this.currentTickers[crypto] = null;
    });
  }

  async fetchAvailableUSDPairs() {
    try {
      console.log('Fetching available USD trading pairs from Coinbase...');
      // Use Coinbase public endpoint to get all products
      const response = await fetch('https://api.coinbase.com/api/v3/brokerage/market/products');

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }

      const data = await response.json();

      // Filter for USD pairs only and active products
      const usdPairs = data.products
        .filter(product =>
          product.quote_currency_id === 'USD' &&
          product.status === 'online' &&
          !product.trading_disabled
        )
        .map(product => product.product_id);

      console.log(`Found ${usdPairs.length} active USD trading pairs`);
      return usdPairs;
    } catch (error) {
      console.error('Failed to fetch USD pairs:', error);
      // Return empty array if API call fails - no hardcoded fallbacks
      console.log('No fallback pairs configured - will retry fetching on next connection');
      return [];
    }
  }

  async initialize() {
    try {
      let finalPairs = [];

      // Handle different crypto configuration types
      if (Array.isArray(this.config.cryptoConfig)) {
        // Custom pairs specified
        finalPairs = this.config.cryptoConfig;
        console.log(`Using custom pairs: ${finalPairs.join(', ')}`);
      } else if (this.config.cryptoConfig && this.config.cryptoConfig.topN) {
        // Top N pairs requested
        const allPairs = await this.fetchAvailableUSDPairs();
        if (allPairs.length > 0) {
          finalPairs = allPairs.slice(0, this.config.cryptoConfig.topN);
          console.log(`Using top ${this.config.cryptoConfig.topN} pairs from ${allPairs.length} available pairs`);
        }
      } else {
        // Default: fetch all available USD pairs
        finalPairs = await this.fetchAvailableUSDPairs();
        console.log(`Using all ${finalPairs.length} available USD pairs`);
      }

      // Update config with final pairs
      this.config.cryptos = finalPairs;

      // Initialize volume tracking for all pairs
      finalPairs.forEach((pair) => {
        this.volumeWindows[pair] = [];
        this.historicalAvgs[pair] = 0;
        this.currentTickers[pair] = null;
      });

      console.log(`Initialized tracking for ${finalPairs.length} USD pairs`);

      // Only connect if we have pairs to monitor
      if (finalPairs.length > 0) {
        this.connect();
      } else {
        console.warn('No USD pairs to monitor - skipping WebSocket connection');
      }
    } catch (error) {
      console.error('Initialization failed:', error);
      // Don't fallback to connect if we have no pairs to monitor
    }
  }

  connect() {
    try {
      this.ws = new WebSocket(this.config.wsUrl);

      this.ws.on("open", () => {
        console.log("WebSocket connected to Coinbase Advanced Trade");
        this.reconnectAttempts = 0;
        this.subscribe();
        this.startTokenRefresh();
      });

      this.ws.on("message", (data) => {
        this.handleMessage(data);
      });

      this.ws.on("error", (error) => {
        console.error("Advanced Trade WebSocket error:", error);
        if (error.message && error.message.includes('401')) {
          console.error('Authentication failed. Please check your COINBASE_API_KEY and COINBASE_SIGNING_KEY.');
        }
      });

      this.ws.on("close", () => {
        console.log("WebSocket disconnected");
        this.stopTokenRefresh();
        this.handleReconnect();
      });
    } catch (error) {
      console.error("Failed to create WebSocket:", error);
      this.handleReconnect();
    }
  }

  async subscribe() {
    try {
      console.log('Starting subscription process...');

      // First, subscribe to heartbeats to keep connection alive
      try {
        const heartbeatJWT = this.jwtManager.getValidToken();
        if (!heartbeatJWT) {
          throw new Error('Failed to generate JWT token for heartbeats');
        }

        const heartbeatMsg = {
          type: "subscribe",
          channel: "heartbeats",
          jwt: heartbeatJWT
        };

        console.log('Subscribing to heartbeats channel...');
        this.ws.send(JSON.stringify(heartbeatMsg));
        console.log('Heartbeats subscription sent successfully');
      } catch (error) {
        console.error('Failed to subscribe to heartbeats:', error.message);
        return; // Don't proceed if JWT generation fails
      }

      // Wait a bit before other subscriptions
      await new Promise(resolve => setTimeout(resolve, 500));

      // Batch products - use smaller batch size (50) for better reliability
      const batchSize = 50;
      const productBatches = [];

      for (let i = 0; i < this.config.cryptos.length; i += batchSize) {
        productBatches.push(this.config.cryptos.slice(i, i + batchSize));
      }

      console.log(`Subscribing to ${this.config.cryptos.length} USD pairs across ${productBatches.length} batches (${batchSize} products per batch)`);

      // Subscribe to each batch
      for (let batchIndex = 0; batchIndex < productBatches.length; batchIndex++) {
        const batch = productBatches[batchIndex];
        console.log(`Processing batch ${batchIndex + 1}/${productBatches.length} with ${batch.length} products`);

        try {
          // Subscribe to ticker_batch (updates every 5 seconds, better for many products)
          const tickerJWT = this.jwtManager.getValidToken();
          if (!tickerJWT) {
            throw new Error('Failed to generate JWT token for ticker_batch');
          }

          const tickerMsg = {
            type: "subscribe",
            product_ids: batch,
            channel: "ticker_batch",
            jwt: tickerJWT
          };

          console.log(`Subscribing to ticker_batch for batch ${batchIndex + 1}...`);
          this.ws.send(JSON.stringify(tickerMsg));
          console.log(`Ticker_batch subscription sent for batch ${batchIndex + 1}`);

          // Increased delay between subscriptions for rate limiting
          await new Promise(resolve => setTimeout(resolve, 300));

          // Subscribe to market_trades for volume tracking
          const tradesJWT = this.jwtManager.getValidToken();
          if (!tradesJWT) {
            throw new Error('Failed to generate JWT token for market_trades');
          }

          const tradesMsg = {
            type: "subscribe",
            product_ids: batch,
            channel: "market_trades",
            jwt: tradesJWT
          };

          console.log(`Subscribing to market_trades for batch ${batchIndex + 1}...`);
          this.ws.send(JSON.stringify(tradesMsg));
          console.log(`Market_trades subscription sent for batch ${batchIndex + 1}`);

          // Longer delay before next batch to respect rate limits
          if (batchIndex < productBatches.length - 1) {
            console.log(`Waiting before next batch (${batchIndex + 2}/${productBatches.length})...`);
            await new Promise(resolve => setTimeout(resolve, 500));
          }

        } catch (error) {
          console.error(`Failed to subscribe to batch ${batchIndex + 1}:`, error.message);
          // Continue with next batch even if this one fails
        }
      }

      console.log("Subscription process completed for all batches");
    } catch (error) {
      console.error('Subscription error:', error);
    }
  }

  handleMessage(data) {
    try {
      const message = JSON.parse(data);

      // Handle messages based on channel (Advanced Trade API uses 'channel' not 'type' at root)
      switch (message.channel) {
        case "ticker":
          // Single ticker update
          if (message.events) {
            message.events.forEach(event => {
              if (event.tickers) {
                event.tickers.forEach(ticker => this.handleTickerMessage(ticker));
              }
            });
          }
          break;
        case "ticker_batch":
          // Batch ticker updates
          if (message.events) {
            message.events.forEach(event => {
              if (event.tickers) {
                event.tickers.forEach(ticker => this.handleTickerMessage(ticker));
              }
            });
          }
          break;
        case "market_trades":
          // Market trades for volume tracking
          if (message.events) {
            message.events.forEach(event => {
              if (event.trades) {
                event.trades.forEach(trade => {
                  const crypto = trade.product_id;
                  const size = parseFloat(trade.size);
                  this.processVolumeData(crypto, size);
                });
              }
            });
          }
          break;
        case "subscriptions":
          console.log("✅ Subscription confirmed:", message.events ? JSON.stringify(message.events[0]).substring(0, 200) : 'OK');
          break;
        case "heartbeats":
          // Heartbeat - connection is alive
          break;
        default:
          // Check if this is an error message
          if (message.type === "error") {
            console.error("❌ Coinbase Advanced Trade error:", message);
            if (message.message) {
              const errorMsg = message.message.toLowerCase();
              if (errorMsg.includes('jwt') || errorMsg.includes('token') || errorMsg.includes('auth')) {
                console.log('JWT authentication failed, refreshing token and resubscribing...');
                this.jwtManager.refreshToken();
                setTimeout(() => {
                  if (this.ws && this.ws.readyState === WebSocket.OPEN) {
                    this.subscribe();
                  }
                }, 1000);
              } else if (errorMsg.includes('unauthorized')) {
                console.error('Authentication failed. Please verify your API credentials.');
              }
            }
          }
          break;
      }
    } catch (error) {
      console.error("Error parsing message:", error);
    }
  }

  handleTickerMessage(ticker) {
    // Update current ticker data - handle Advanced Trade API format
    const crypto = ticker.product_id;
    const price = parseFloat(ticker.price);
    const volume_24h = parseFloat(ticker.volume_24_h || ticker.volume_24h || 0);
    const low_24h = parseFloat(ticker.low_24_h || ticker.low_24h || 0);
    const high_24h = parseFloat(ticker.high_24_h || ticker.high_24h || 0);

    // Calculate open_24h from price_percent_chg_24_h if available
    const price_percent_chg = parseFloat(ticker.price_percent_chg_24_h || 0);
    const open_24h = price_percent_chg !== 0 ? price / (1 + price_percent_chg / 100) : price;

    this.currentTickers[crypto] = {
      crypto: crypto,
      price: price,
      bid: parseFloat(ticker.best_bid || ticker.bid || 0),
      ask: parseFloat(ticker.best_ask || ticker.ask || 0),
      volume_24h: volume_24h,
      price_24h: open_24h,
      low_24h: low_24h,
      high_24h: high_24h,
      price_change_24h: price - open_24h,
      price_change_percent_24h: price_percent_chg,
      time: ticker.time || ticker.timestamp || new Date().toISOString(),
      sequence: ticker.sequence || Date.now(),
    };

    // Emit ticker update for frontend
    this.emit("ticker_update", this.currentTickers[crypto]);

    // Emit priority update if this is a priority pair
    if (this.priorityPairs.has(crypto)) {
      const now = Date.now();
      const lastUpdate = this.lastPriorityUpdate[crypto] || 0;

      // Emit priority updates more frequently
      if (now - lastUpdate >= this.priorityUpdateInterval) {
        this.emit("priority_ticker_update", this.currentTickers[crypto]);
        this.lastPriorityUpdate[crypto] = now;
      }
    }
  }

  handleAdvancedTradeMessage(event, channel) {
    if (channel === "market_trades" && event.trades) {
      event.trades.forEach(trade => {
        const crypto = trade.product_id;
        const size = parseFloat(trade.size);
        this.processVolumeData(crypto, size);
      });
    } else if (channel === "ticker" && event.tickers) {
      event.tickers.forEach(ticker => {
        this.handleAdvancedTradeTicker(ticker);
      });
    }
  }

  handleAdvancedTradeTicker(ticker) {
    // Convert Advanced Trade ticker format to standardized format
    const crypto = ticker.product_id;
    const standardFormat = {
      product_id: crypto,
      price: ticker.price,
      best_bid: ticker.best_bid,
      best_ask: ticker.best_ask,
      volume_24h: ticker.volume_24h,
      open_24h: ticker.open_24h,
      low_24h: ticker.low_24h,
      high_24h: ticker.high_24h,
      time: ticker.time,
      sequence: ticker.sequence || Date.now(), // Fallback if sequence not provided
    };

    this.handleTickerMessage(standardFormat);
  }

  processVolumeData(crypto, size) {
    const now = Date.now();
    const windowMs = this.config.windowMinutes * 60 * 1000;

    // Add to volume window
    this.volumeWindows[crypto].push({
      time: now,
      size: size,
    });

    // Remove old entries outside the window
    this.volumeWindows[crypto] = this.volumeWindows[crypto].filter(
      (entry) => entry.time > now - windowMs
    );

    // Calculate current window volume
    const currentVolume = this.volumeWindows[crypto].reduce(
      (sum, entry) => sum + entry.size,
      0
    );

    // Check for volume surge
    if (this.historicalAvgs[crypto] > 0) {
      const ratio = currentVolume / this.historicalAvgs[crypto];

      if (ratio > this.config.volumeThreshold) {
        const alert = {
          crypto: crypto,
          current_vol: currentVolume,
          avg_vol: this.historicalAvgs[crypto],
          threshold: this.config.volumeThreshold,
          ratio: ratio,
          ticker: this.currentTickers[crypto], // Include current ticker data
        };

        this.emit("volume_alert", alert);

        // Update historical average (simple moving average)
        this.historicalAvgs[crypto] =
          this.historicalAvgs[crypto] * 0.9 + currentVolume * 0.1;
      }
    } else {
      // Initialize historical average if not set
      this.historicalAvgs[crypto] = currentVolume;
    }
  }

  handleReconnect() {
    const delay = Math.min(
      this.config.reconnectDelay * Math.pow(2, this.reconnectAttempts),
      this.config.maxReconnectDelay
    );

    console.log(`Reconnecting in ${delay}ms...`);
    this.reconnectAttempts++;

    // Refresh JWT token before reconnecting
    this.jwtManager.refreshToken();

    setTimeout(() => {
      this.connect();
    }, delay);
  }

  getCurrentTicker(crypto) {
    return this.currentTickers[crypto];
  }

  getAllTickers() {
    return this.currentTickers;
  }

  startTokenRefresh() {
    // Clear any existing interval
    this.stopTokenRefresh();

    // Refresh token every 90 seconds (30 seconds before expiry)
    this.tokenRefreshInterval = setInterval(() => {
      if (this.jwtManager.isTokenExpired()) {
        console.log('JWT token expiring soon, refreshing proactively...');
        this.jwtManager.refreshToken();

        // Resubscribe with new token
        if (this.ws && this.ws.readyState === WebSocket.OPEN) {
          this.subscribe();
        }
      }
    }, 90000); // 90 seconds
  }

  stopTokenRefresh() {
    if (this.tokenRefreshInterval) {
      clearInterval(this.tokenRefreshInterval);
      this.tokenRefreshInterval = null;
    }
  }

  disconnect() {
    this.stopTokenRefresh();
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }

  // Priority system methods for active trading pairs
  addPriorityPair(productId) {
    this.priorityPairs.add(productId);
    this.lastPriorityUpdate[productId] = 0; // Reset update timer
    console.log(`Added ${productId} to priority monitoring. Total priority pairs: ${this.priorityPairs.size}`);
  }

  removePriorityPair(productId) {
    this.priorityPairs.delete(productId);
    delete this.lastPriorityUpdate[productId];
    console.log(`Removed ${productId} from priority monitoring. Total priority pairs: ${this.priorityPairs.size}`);
  }

  getPriorityPairs() {
    return Array.from(this.priorityPairs);
  }

  isPriorityPair(productId) {
    return this.priorityPairs.has(productId);
  }

  clearPriorityPairs() {
    this.priorityPairs.clear();
    this.lastPriorityUpdate = {};
    console.log('Cleared all priority pairs');
  }

  // Get priority statistics
  getPriorityStats() {
    return {
      totalPriorityPairs: this.priorityPairs.size,
      priorityPairs: Array.from(this.priorityPairs),
      updateInterval: this.priorityUpdateInterval,
      lastUpdates: { ...this.lastPriorityUpdate }
    };
  }
}

module.exports = CoinbaseWebSocketHandler;

File: websocket-service/websocket_handler.py
#!/usr/bin/env python3
"""
Enhanced WebSocket Handler - Drop-in replacement for existing websocket-handler.js
Provides the same interface with multi-connection reliability
"""

import asyncio
import json
import logging
import time
import threading
from typing import Dict, List, Optional, Callable, Any
from datetime import datetime, timedelta
import requests
from multi_ws_manager import MultiWSManager, TickerData, ConnectionStats

logger = logging.getLogger(__name__)

class EnhancedWebSocketHandler:
    """
    Enhanced WebSocket Handler with multi-connection support
    Drop-in replacement for the existing Node.js handler
    """

    def __init__(self, config: dict):
        self.config = {
            'wsUrl': config.get('wsUrl', 'wss://advanced-trade-ws.coinbase.com'),
            'cryptoConfig': config.get('cryptoConfig', None),
            'volumeThreshold': config.get('volumeThreshold', 1.5),
            'windowMinutes': config.get('windowMinutes', 5),
            'productsPerConnection': config.get('productsPerConnection', 15),
            'apiKey': config.get('apiKey'),
            'signingKey': config.get('signingKey')
        }

        # Data storage (same as original)
        self.currentTickers: Dict[str, Dict] = {}
        self.volumeWindows: Dict[str, List] = {}
        self.historicalAvgs: Dict[str, float] = {}

        # Priority system (same as original)
        self.priorityPairs = set()
        self.priorityUpdateInterval = 1000
        self.lastPriorityUpdate = {}

        # Multi-connection manager
        self.ws_manager: Optional[MultiWSManager] = None

        # Event callbacks (same interface as EventEmitter)
        self.event_callbacks: Dict[str, List[Callable]] = {}

        # Initialize flag
        self.initialized = False

        # 24-hour baseline tracking
        self.baseline_prices: Dict[str, float] = {}  # Store baseline prices from midnight
        self.last_baseline_update = None
        self.baseline_thread = None
        self.running = False

    def on(self, event: str, callback: Callable):
        """Add event listener (same interface as EventEmitter)"""
        if event not in self.event_callbacks:
            self.event_callbacks[event] = []
        self.event_callbacks[event].append(callback)

    def emit(self, event: str, data: Any):
        """Emit event to all listeners"""
        if event in self.event_callbacks:
            for callback in self.event_callbacks[event]:
                try:
                    callback(data)
                except Exception as e:
                    logger.error(f"Error in event callback for {event}: {e}")

    async def fetchAvailableUSDPairs(self) -> List[str]:
        """Fetch available USD trading pairs (async version)"""
        try:
            logger.info('Fetching available USD trading pairs from Coinbase...')
            response = requests.get(
                'https://api.coinbase.com/api/v3/brokerage/market/products',
                timeout=10
            )
            response.raise_for_status()

            data = response.json()

            # Filter for USD pairs only and active products
            usd_pairs = [
                product['product_id']
                for product in data.get('products', [])
                if (product.get('quote_currency_id') == 'USD' and
                    product.get('status') == 'online' and
                    not product.get('trading_disabled', False))
            ]

            logger.info(f'Found {len(usd_pairs)} active USD trading pairs')
            return usd_pairs

        except Exception as e:
            logger.error(f'Failed to fetch USD pairs: {e}')
            return []

    async def initialize(self):
        """Initialize the enhanced WebSocket handler"""
        try:
            final_pairs = []

            # Handle different crypto configuration types (same logic as original)
            if isinstance(self.config['cryptoConfig'], list):
                # Custom pairs specified
                final_pairs = self.config['cryptoConfig']
                logger.info(f'Using custom pairs: {", ".join(final_pairs)}')

            elif (self.config['cryptoConfig'] and
                  isinstance(self.config['cryptoConfig'], dict) and
                  'topN' in self.config['cryptoConfig']):
                # Top N pairs requested
                all_pairs = await self.fetchAvailableUSDPairs()
                if all_pairs:
                    final_pairs = all_pairs[:self.config['cryptoConfig']['topN']]
                    logger.info(f'Using top {self.config["cryptoConfig"]["topN"]} pairs from {len(all_pairs)} available pairs')

            else:
                # Default: fetch all available USD pairs
                final_pairs = await self.fetchAvailableUSDPairs()
                logger.info(f'Using all {len(final_pairs)} available USD pairs')

            if not final_pairs:
                logger.warning('No USD pairs to monitor - initialization failed')
                return

            # Initialize volume tracking for all pairs
            for pair in final_pairs:
                self.volumeWindows[pair] = []
                self.historicalAvgs[pair] = 0
                self.currentTickers[pair] = None

            # Setup multi-connection WebSocket manager
            self.ws_manager = MultiWSManager(
                products=final_pairs,
                products_per_connection=self.config['productsPerConnection']
            )

            # Add callbacks
            self.ws_manager.add_data_callback(self._handle_ticker_update)
            self.ws_manager.add_stats_callback(self._handle_connection_stats)

            logger.info(f'Enhanced WebSocket handler initialized for {len(final_pairs)} pairs')
            self.initialized = True

            # Start connections
            self.ws_manager.start()

            # Start baseline update thread (runs at midnight daily)
            self.running = True
            self.baseline_thread = threading.Thread(target=self._baseline_update_loop, daemon=True)
            self.baseline_thread.start()
            logger.info('Started 24hr baseline price update thread (runs at midnight)')

            # Fetch initial baseline prices after a short delay (to let tickers populate)
            threading.Timer(5.0, self._fetch_baseline_prices).start()

        except Exception as e:
            logger.error(f'Enhanced WebSocket handler initialization failed: {e}')
            raise

    def _handle_ticker_update(self, ticker_data: TickerData):
        """Handle ticker updates from multi-connection manager"""
        try:
            # Calculate 24hr change from baseline (midnight price)
            baseline_price = self.baseline_prices.get(ticker_data.product_id, ticker_data.price)
            price_change_24h = ticker_data.price - baseline_price
            price_change_percent_24h = ((ticker_data.price - baseline_price) / baseline_price * 100) if baseline_price > 0 else 0

            # Convert to original format
            ticker_dict = {
                'crypto': ticker_data.product_id,
                'price': ticker_data.price,
                'bid': ticker_data.best_bid or 0,
                'ask': ticker_data.best_ask or 0,
                'volume_24h': ticker_data.volume_24h or 0,
                'price_24h': baseline_price,  # Baseline from midnight
                'low_24h': 0,    # Not available in real-time ticker
                'high_24h': 0,   # Not available in real-time ticker
                'price_change_24h': price_change_24h,
                'price_change_percent_24h': price_change_percent_24h,
                'time': ticker_data.time,
                'sequence': ticker_data.sequence,
                'last_size': ticker_data.last_size,
                'timestamp': datetime.now().strftime('%H:%M:%S')
            }

            # Store in current tickers
            self.currentTickers[ticker_data.product_id] = ticker_dict

            # Emit ticker_update event (same as original)
            self.emit('ticker_update', ticker_dict)

            # Handle priority pairs
            if ticker_data.product_id in self.priorityPairs:
                now = time.time() * 1000  # Convert to milliseconds
                last_update = self.lastPriorityUpdate.get(ticker_data.product_id, 0)

                if now - last_update >= self.priorityUpdateInterval:
                    self.emit('priority_ticker_update', ticker_dict)
                    self.lastPriorityUpdate[ticker_data.product_id] = now

            # Process volume data if last_size available
            if ticker_data.last_size > 0:
                self._processVolumeData(ticker_data.product_id, ticker_data.last_size)

        except Exception as e:
            logger.error(f'Error handling ticker update: {e}')

    def _handle_connection_stats(self, stats: ConnectionStats):
        """Handle connection statistics"""
        # You can emit connection health events here if needed
        pass

    def _processVolumeData(self, crypto: str, size: float):
        """Process volume data (same logic as original)"""
        now = time.time() * 1000  # Convert to milliseconds
        window_ms = self.config['windowMinutes'] * 60 * 1000

        # Add to volume window
        self.volumeWindows[crypto].append({
            'time': now,
            'size': size
        })

        # Remove old entries outside the window
        self.volumeWindows[crypto] = [
            entry for entry in self.volumeWindows[crypto]
            if entry['time'] > now - window_ms
        ]

        # Calculate current window volume
        current_volume = sum(entry['size'] for entry in self.volumeWindows[crypto])

        # Check for volume surge
        if self.historicalAvgs[crypto] > 0:
            ratio = current_volume / self.historicalAvgs[crypto]

            if ratio > self.config['volumeThreshold']:
                alert = {
                    'crypto': crypto,
                    'current_vol': current_volume,
                    'avg_vol': self.historicalAvgs[crypto],
                    'threshold': self.config['volumeThreshold'],
                    'ratio': ratio,
                    'ticker': self.currentTickers.get(crypto)
                }

                self.emit('volume_alert', alert)

                # Update historical average (simple moving average)
                self.historicalAvgs[crypto] = (
                    self.historicalAvgs[crypto] * 0.9 + current_volume * 0.1
                )
        else:
            # Initialize historical average if not set
            self.historicalAvgs[crypto] = current_volume

    def getCurrentTicker(self, crypto: str) -> Optional[Dict]:
        """Get current ticker for a crypto (same interface as original)"""
        return self.currentTickers.get(crypto)

    def getAllTickers(self) -> Dict[str, Dict]:
        """Get all current tickers (same interface as original)"""
        return self.currentTickers.copy()

    def disconnect(self):
        """Disconnect all WebSocket connections"""
        self.running = False
        if self.ws_manager:
            self.ws_manager.stop()

    def _fetch_baseline_prices(self):
        """Fetch current prices from all tracked pairs to use as baseline"""
        try:
            # Use current websocket prices as baseline (they're real-time and accurate)
            baseline_count = 0
            for product_id, ticker in self.currentTickers.items():
                if 'price' in ticker and ticker['price'] > 0:
                    self.baseline_prices[product_id] = ticker['price']
                    baseline_count += 1

            self.last_baseline_update = datetime.now()
            logger.info(f"✅ Baseline prices updated: {baseline_count} products at {self.last_baseline_update.strftime('%Y-%m-%d %H:%M:%S')}")

        except Exception as e:
            logger.error(f"Error fetching baseline prices: {e}")

    def _baseline_update_loop(self):
        """Background thread that updates baseline at midnight every day"""
        while self.running:
            try:
                now = datetime.now()

                # Calculate next midnight
                tomorrow = now + timedelta(days=1)
                next_midnight = datetime(tomorrow.year, tomorrow.month, tomorrow.day, 0, 0, 0)

                # Calculate seconds until midnight
                seconds_until_midnight = (next_midnight - now).total_seconds()

                logger.info(f"Next baseline update scheduled for: {next_midnight.strftime('%Y-%m-%d %H:%M:%S')} ({seconds_until_midnight/3600:.1f} hours)")

                # Sleep until midnight
                time.sleep(seconds_until_midnight)

                # Update baseline prices at midnight
                if self.running:
                    logger.info("🕛 Midnight reached - updating 24hr baseline prices")
                    self._fetch_baseline_prices()

            except Exception as e:
                logger.error(f"Error in baseline update loop: {e}")
                # Sleep for an hour and try again if there's an error
                time.sleep(3600)

    # Priority pair management methods (same interface as original)
    def addPriorityPair(self, product_id: str):
        """Add a priority pair"""
        self.priorityPairs.add(product_id)
        self.lastPriorityUpdate[product_id] = 0
        logger.info(f'Added {product_id} to priority monitoring. Total priority pairs: {len(self.priorityPairs)}')

    def removePriorityPair(self, product_id: str):
        """Remove a priority pair"""
        self.priorityPairs.discard(product_id)
        self.lastPriorityUpdate.pop(product_id, None)
        logger.info(f'Removed {product_id} from priority monitoring. Total priority pairs: {len(self.priorityPairs)}')

    def getPriorityPairs(self) -> List[str]:
        """Get list of priority pairs"""
        return list(self.priorityPairs)

    def isPriorityPair(self, product_id: str) -> bool:
        """Check if product is a priority pair"""
        return product_id in self.priorityPairs

    def clearPriorityPairs(self):
        """Clear all priority pairs"""
        self.priorityPairs.clear()
        self.lastPriorityUpdate.clear()
        logger.info('Cleared all priority pairs')

    def getPriorityStats(self) -> Dict:
        """Get priority statistics"""
        return {
            'totalPriorityPairs': len(self.priorityPairs),
            'priorityPairs': list(self.priorityPairs),
            'updateInterval': self.priorityUpdateInterval,
            'lastUpdates': self.lastPriorityUpdate.copy()
        }

    def getHealthSummary(self) -> Dict:
        """Get health summary of the WebSocket connections"""
        if not self.ws_manager:
            return {'status': 'not_initialized'}

        return self.ws_manager.get_health_summary()


# Factory function to create enhanced handler (same interface as original)
def create_websocket_handler(config: dict) -> EnhancedWebSocketHandler:
    """Create an enhanced WebSocket handler"""
    return EnhancedWebSocketHandler(config)